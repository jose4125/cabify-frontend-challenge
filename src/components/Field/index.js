import React from 'react';

const Field = ({
  value,
  type,
  name,
  stateName,
  onHandleFocus,
  onHandleBlur,
  onHandleChange
}) => {
  return (
    <input
      autoComplete="off"
      type={type}
      name={name}
      value={value}
      onChange={event => onHandleChange(event.target.value, stateName)}
      onFocus={() => onHandleFocus(stateName)}
      onBlur={() => onHandleBlur(stateName)}
    />
  );
};

export default Field;
