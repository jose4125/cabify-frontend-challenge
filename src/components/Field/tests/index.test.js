import React from 'react';
import { shallow } from 'enzyme';

import Field from '../index';

describe('Field', () => {
  const mockOnHandleChange = jest.fn();
  const mockOnHandleFocus = jest.fn();
  const mockOnHandleBlur = jest.fn();
  const context = {
    state: {
      fullName: '',
      jobDescription: '',
      prefix: '',
      phoneNumber: '',
      email: '',
      website: '',
      address: ''
    }
  };

  const props = {
    type: 'text',
    name: 'fullname',
    value: '',
    stateName: 'fullName',
    onHandleFocus: mockOnHandleFocus,
    onHandleBlur: mockOnHandleBlur,
    onHandleChange: mockOnHandleChange
  };

  const wrapper = shallow(<Field {...props} />);

  it('should render with out crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have an input tag', () => {
    expect(wrapper.find('input').exists()).toBe(true);
  });

  it('should the input have a type', () => {
    expect(wrapper.find('input').prop('type')).toBe(props.type);
  });

  it('should the input have a name', () => {
    expect(wrapper.find('input').prop('name')).toBe(props.name);
  });

  it('should the input have an empty value', () => {
    expect(wrapper.find('input').prop('value')).toBe('');
  });

  it('should the input call onHandleFocus', () => {
    const fullName = 'john doe';
    wrapper.find('input').simulate('focus');
    expect(mockOnHandleFocus).toHaveBeenCalled();
  });

  it('should the input call handleChange', () => {
    const fullName = 'john doe';
    wrapper.find('input').simulate('change', { target: { value: fullName } });
    expect(mockOnHandleChange).toHaveBeenCalledWith(fullName, props.stateName);
  });

  it('should the input call onHandleBlur with stateName', () => {
    wrapper.find('input').simulate('blur');
    expect(mockOnHandleBlur).toHaveBeenCalledWith(props.stateName);
  });

  describe('update context Field', () => {
    const props = {
      type: 'text',
      name: 'fullname',
      stateName: 'fullName',
      value: 'john doe',
      onHandleFocus: mockOnHandleFocus,
      onHandleBlur: mockOnHandleBlur,
      onHandleChange: mockOnHandleChange
    };

    const wrapper = shallow(<Field {...props} />);

    it('should the input have a value', () => {
      expect(wrapper.find('input').prop('value')).toBe(props.value);
    });
  });
});
