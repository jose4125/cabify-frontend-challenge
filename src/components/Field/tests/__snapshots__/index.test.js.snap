// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field should render with out crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Field
    name="fullname"
    onHandleBlur={[MockFunction]}
    onHandleChange={[MockFunction]}
    onHandleFocus={[MockFunction]}
    stateName="fullName"
    type="text"
    value=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "autoComplete": "off",
      "name": "fullname",
      "onBlur": [Function],
      "onChange": [Function],
      "onFocus": [Function],
      "type": "text",
      "value": "",
    },
    "ref": null,
    "rendered": null,
    "type": "input",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "autoComplete": "off",
        "name": "fullname",
        "onBlur": [Function],
        "onChange": [Function],
        "onFocus": [Function],
        "type": "text",
        "value": "",
      },
      "ref": null,
      "rendered": null,
      "type": "input",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": false,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
